src/main/java/org/opensearch/flowframework/workflow/ReIndexStep.java

```diff
@@ -0,0 +1,175 @@
+import org.opensearch.common.Booleans;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.flowframework.exception.FlowFrameworkException;
+import org.opensearch.flowframework.exception.WorkflowStepException;
+import org.opensearch.flowframework.indices.FlowFrameworkIndicesHandler;
+import org.opensearch.flowframework.util.ParseUtils;
+import org.opensearch.index.reindex.BulkByScrollResponse;
+import org.opensearch.index.reindex.ReindexAction;
+import org.opensearch.index.reindex.ReindexRequest;
+
+import java.util.Map;
+import java.util.Set;
+
+import static org.opensearch.flowframework.common.CommonValue.DESTINATION_INDEX;
+import static org.opensearch.flowframework.common.CommonValue.MAX_DOCS;
+import static org.opensearch.flowframework.common.CommonValue.REFRESH;
+import static org.opensearch.flowframework.common.CommonValue.REQUESTS_PER_SECOND;
+import static org.opensearch.flowframework.common.CommonValue.REQUIRE_ALIAS;
+import static org.opensearch.flowframework.common.CommonValue.RE_INDEX_FIELD;
+import static org.opensearch.flowframework.common.CommonValue.SLICES;
+import static org.opensearch.flowframework.common.CommonValue.SOURCE_INDEX;
+
+/**
+ * Step to reindex
+ */
+public class ReIndexStep implements WorkflowStep {
+
+    private static final Logger logger = LogManager.getLogger(ReIndexStep.class);
+    private final Client client;
+    private final FlowFrameworkIndicesHandler flowFrameworkIndicesHandler;
+
+    /** The name of this step, used as a key in the template and the {@link WorkflowStepFactory} */
+    public static final String NAME = "reindex";
+
+    /**
+     * Instantiate this class
+     *
+     * @param client Client to create an index
+     * @param flowFrameworkIndicesHandler FlowFrameworkIndicesHandler class to update system indices
+     */
+    public ReIndexStep(Client client, FlowFrameworkIndicesHandler flowFrameworkIndicesHandler) {
+        this.client = client;
+        this.flowFrameworkIndicesHandler = flowFrameworkIndicesHandler;
+    }
+
+    @Override
+    public PlainActionFuture<WorkflowData> execute(
+        String currentNodeId,
+        WorkflowData currentNodeInputs,
+        Map<String, WorkflowData> outputs,
+        Map<String, String> previousNodeInputs,
+        Map<String, String> params
+    ) {
+
+        PlainActionFuture<WorkflowData> reIndexFuture = PlainActionFuture.newFuture();
+
+        Set<String> requiredKeys = Set.of(SOURCE_INDEX, DESTINATION_INDEX);
+
+        Set<String> optionalKeys = Set.of(REFRESH, REQUESTS_PER_SECOND, REQUIRE_ALIAS, SLICES, MAX_DOCS);
+
+        try {
+            Map<String, Object> inputs = ParseUtils.getInputsFromPreviousSteps(
+                requiredKeys,
+                optionalKeys,
+                currentNodeInputs,
+                outputs,
+                previousNodeInputs,
+                params
+            );
+
+            String sourceIndices = (String) inputs.get(SOURCE_INDEX);
```

Comment:
Should this be a `String[]`?

Reply:
The list of indices is under the same string.

Reply:
`String ...` varargs is syntactic sugar for `String[]`.  And the internal parameter that gets set is a `String[]`.  Passing a single String in this line as you do here would turn it into a 1-element String array.  

I already approved if we're only allowing reindexing one index, but this seems something we should handle.

Reply:
Looks pretty easy to support a comma-delimited list of indices with [`Strings.splitStringByCommaToArray((String) value)`](https://github.com/opensearch-project/OpenSearch/blob/2e497436cf9dd78aa3d06f86285aef897c962cac/server/src/main/java/org/opensearch/index/reindex/ReindexRequest.java#L395-L409)

Reply:
Pushed the change